File: leetcode/0055-jump-game.cc
// https://leetcode.com/problems/jump-game/
// Medium

#include<vector>
#include<iostream>
using namespace std;

class Solution {
public:
    bool canJump(vector<int>& nums) {
        int maxSoFar = nums[0], size = nums.size();
        for (int i = 0; i < size; ++i) {
            if (maxSoFar >= size-1) return true;
            if (i > maxSoFar) return false;
            maxSoFar = maxSoFar < i + nums[i] ? i + nums[i] : maxSoFar;
        }
        return false;
    }
};

File: leetcode/python/0055-jump-game.py
# https://leetcode.com/problems/jump-game/
# Medium

# Just have a variable that keeps track of the max jump reachable
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        maxJump = 0
        for i,n in enumerate(nums):
            if i > maxJump:
                return False
            maxJump = max(i+n, maxJump)
        return True